Assets/Scripts/Controllers/BattlePrepController.cs

```diff
@@ -0,0 +1,179 @@
+            await PepemonMatchmaker.Enter(selectedLeague, selectedDeck);
+        }
+        catch (Exception e)
+        {
+            Debug.Log("Unable to enter the matchmaker: " + e);
+            failedToEnter = true;
+        }
+
+        // check if the current player's deck is in the matchmaker list of deck owners, if it is, then we can assume the player
+        // is in the waitlist
+        var deckOwner = await PepemonMatchmaker.GetDeckOwner(selectedLeague, selectedDeck);
+
+        string player1addr = Web3Controller.instance.SelectedAccountAddress, 
+               player2addr = null;
+
+        // If the player is in the waitlist, then its because a battle has *not* started yet, the current player's address will
+        // be in the second parameter of the BattleCreated event in this case.
+        // However if the battle *has* started, the player's address will be in the first parameter of the BattleCreated event
+        if (deckOwner == Web3Controller.instance.SelectedAccountAddress)
+        {
+            player2addr = player1addr;
+            player1addr = null;
+            // the Exit button must only be enabled once we are sure that the current player is in the wait list of the Matchmaker
+            _exitButton.GetComponent<Button>().interactable = true;
+        }
+        else if (failedToEnter)
+        {
+            // go back to deck selection since the player could not enter nor is already waiting
+            FindObjectOfType<MainMenuController>().ShowScreen(4);
+            return;
+        }
+
+        Debug.Log("Waiting for CreatedBattle events");
+
+        _cancellationTokenSource = new CancellationTokenSource();
+        var battleEvent = await PepemonBattle.WaitForCreatedBattle(
+                player1addr,
+                player2addr,
+                nextBlock, 
+                _cancellationTokenSource.Token);
+
+        // initiate the battle immediatelly if the player initiated the battle
+        if (battleEvent != null)
+        {
+            await InitBattleScene((PepemonBattle.BattleCreatedEventData) battleEvent);
+        }
+    }
+
+    private async Task InitBattleScene(PepemonBattle.BattleCreatedEventData battleEventData)
+    {
+        Debug.Log("Received battle event. BattleId: " + battleEventData.BattleId);
+        Debug.Log("Loading battle data..");
+
+        var reqBattleRngSeed = PepemonBattle.GetBattleRNGSeed(battleEventData.BattleId);
+        var reqPlayer1BattleCard = PepemonCardDeck.GetBattleCard(battleEventData.Player1Deck);
+        var reqPlayer2BattleCard = PepemonCardDeck.GetBattleCard(battleEventData.Player2Deck);
+        var reqPlayer1SupportCards = PepemonCardDeck.GetAllSupportCards(battleEventData.Player1Deck);
+        var reqPlayer2SupportCards = PepemonCardDeck.GetAllSupportCards(battleEventData.Player2Deck);
+        await Task.WhenAll(reqBattleRngSeed, reqPlayer1BattleCard, reqPlayer2BattleCard, reqPlayer1SupportCards, reqPlayer2SupportCards);
```

Comment:
There's a window here where the battle has been created but the exit button can still be pressed.
I'd recommend disabling input on the exit button after the BattleCreated event is picked up

Reply:
Updated in [this ](https://github.com/pepemon-dao/pepemon-battle-degen/pull/48/commits/60b768f7be47f29f9a22f5590f431c19447d0f3a) commit